src/main/java/com/minecolonies/coremod/colony/ColonyList.java

```diff
@@ -0,0 +1,240 @@
+    }
+
+    /**
+     * Add a new Colony to the List.
+     * @param colony colony to add to the list.
+     */
+    public void add(T colony)
+    {
+        final T existingColony = get(colony.getID());
+        if (existingColony != null && existingColony != colony)
+        {
+            throw new IllegalArgumentException(
+                    String.format("Already a colony registered to id=%d, colony=%s, not changing to colony=%s",
+                            colony.getID(),
+                            existingColony.getName(),
+                            colony.getName()));
+        }
+
+        list[colony.getID()] = colony;
+    }
+
+    /**
+     * Remove the Colony from the list.
+     * @param colony the Colony to remove.
+     */
+    public void remove(T colony)
+    {
+        remove(colony.getID());
+    }
+
+    /**
+     * Remove the colony with the provided id from the list.
+     * @param id colony id to remove.
+     */
+    public void remove(int id)
+    {
+        if (list[id] == null)
+        {
+            Log.getLogger().warn("Tried to remove colony with id=%d, but it didn't exist.", id);
+        }
+
+        size--;
+        list[id] = null;
+
+        if (!nullIndices.contains(id))
+        {
+            nullIndices.add(id);
+        }
+    }
+
+    private int getNextColonyID()
+    {
+        if (nullIndices.isEmpty())
+        {
+            topID++;
+            if (topID >= list.length)
+            {
+                // Expand list
+                Object[] newList = new Object[list.length * 2];
+                System.arraycopy(list, 0, newList, 0, list.length);
+                list = newList;
+            }
+
+            return topID;
+        }
+
+        return nullIndices.remove(0);
+    }
+
+    /**
+     * Empty the list.
+     */
+    public void clear()
+    {
+        for (int i = 0; i < list.length; i++)
+        {
+            list[i] = null;
+        }
+
+        nullIndices.clear();
+
+        topID = 0;
+        size = 0;
+    }
+
+    /**
+     * Return the number of Colonies in the list.
+     * @return number of Colonies in the list.
+     */
+    public int size()
+    {
+        return size;
+    }
+
+    /**
+     * Checks if there are Colonies in the list.
+     * @return true if there are no Colonies.
+     */
+    public boolean isEmpty()
+    {
+        return size == 0;
+    }
+
+    /**
+     * Copy all of the colonies to a List. Because this does a copy, it should only
+     * be used when really needed.
+     * @return List of Colonies.
+     */
+    @NotNull
+    public List<T> getCopyAsList()
+    {
+        final List<T> copyList = new ArrayList<>();
+        for (final T colony : this)
+        {
+            copyList.add(colony);
+        }
+
+        return copyList;
+    }
+
+    /**
+     * Makes an iterator for the list.
+     * @return an iterator for the colonies.
+     */
+    @Override
+    public Iterator<T> iterator()
+    {
+        return new Iterator<T>()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this anonymous class number of lines from 32 to at most 30, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1188)